### PR TITLE
Compare Models Across PLIER Dimensions

### PR DESCRIPTION
**Pull request by ben-heil**
This PR mostly consists of updates to existing scripts to enable different PLIER outputs to be passed in.

Are validation set accuracy and area under the receiver operating curve the best evaluation metrics to use here? Or should I be using something else to compare models?

Changes:
- Created `compare_models.ipynb`, a notebook for visualizing the output of `evaluate_models`
- Added `create_identity_matrix.py`, a script designed to store an identity matrix in a pandas dataframe that is compatible with `evaluate_models.py`
- Modified `evaluate_models.py` to read in multiple files and wrap the model training code in a try/finally block to ensure data is saved if an error is thrown. All the new logic happens in lines 19-35, the other lines are stylistic changes to comply with my new linter
- Created `model_evaluation_pipeline.sh`, a pipeline script to call `run_plier.R` multiple times in parallell, then call `evaulate_models.py` on the results
- `run_plier.R` now takes the output directory and value of `k` (the number of PLIER LVs) as parameters. Also improved PLIER installation code to fix devtools issue

### CODE REVIEW

File: brdnet/create_identity_matrix_df.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    parser = argparse.ArgumentParser()
+    parser.add_argument('plier_file', help="A file output by run_plier.R, typically "
+                                           "named something like 'plier_10_Z.tsv'")
+    parser.add_argument('out_file', help='The path to save the resulting tsv to',
+                        default='../data/plier_out/plier_all_Z.tsv')
+    args = parser.parse_args()
+
+    plier_df = pandas.read_csv(args.plier_file, sep='\t')
+
+    identity_matrix = numpy.identity(len(plier_df.index))
```
Comment (ajlee21) [inline brdnet/create_identity_matrix_df.py:18]: I'm not very familiar with PLIER, just curious what this identity matrix is used for?

Comment (ajlee21) [reply brdnet/create_identity_matrix_df.py:18]: As a followup, it might be helpful to have a short description of what PLIER is and why/how you're using it to evaluate your models.  Maybe you could put this in the readme

File: brdnet/evaluate_models.py [hunk 2]
```diff
@@ -5,17 +5,35 @@
...
 import models
 import utils
 
+# Mute INFO and WARNING logs from tensorflow
+os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'
+
+
+def get_Z_files(Z_dir):
+    '''Get the path to all files in the given directory that look like they were
+    created by run_plier.R
```
Comment (ajlee21) [inline brdnet/evaluate_models.py]: maybe add "(i.e. they have the extension _Z.tsv)"

File: brdnet/evaluate_models.py [hunk 3]
```diff
@@ -96,7 +114,7 @@ def eval_pyod_model(model_name, train_X, train_Y, val_X, val_Y):
     module = importlib.import_module(model_module)
     model = getattr(module, model_name)
 
-    model_instance = model()
+    model_instance = model(contamination=.5)
```
Comment (ajlee21) [inline brdnet/evaluate_models.py:117]: What does this "contamination" parameter do?  Based on the name I would guess it is adding noise to your data.

File: brdnet/evaluate_models.py [hunk 4]
```diff
@@ -201,39 +222,49 @@ def eval_tf_model(model_name, lr, train_X, train_Y, val_X, val_Y, checkpoint_dir
...
-                                                                'val_acc',
-                                                                'val_auroc',
-                                                                'val_baseline',
-                                                               ])
-
-    results_df.to_csv(args.out_path)
+    Z_files = get_Z_files(args.Z_file_dir)
+
+    try:
+        for Z_file_path in Z_files:
```
Comment (ajlee21) [inline brdnet/evaluate_models.py:228]: What is this Z_file?  Is this the embedded latent space representation of the data?  Or is this the encoder NN that embeds the data into a lower dimensional latent space?

File: brdnet/model_evaluation_pipeline.sh [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+# evaluate the results of the number of PLIER variables on the models' performance
+
+# The number of processes you want to run at a time for paralellizing run_plier.R
+NUM_PROCESSES=4
+
+PLIER_OUT='../data/plier_out'
+
+# This array is intentionally reverse sorted for more efficient scheduling (you want
+# the longest jobs to be loaded first)
+K_VALS=(100 50 10)
```
Comment (ajlee21) [inline brdnet/model_evaluation_pipeline.sh:12]: What are these K_vals?  How are they used?

File: brdnet/run_plier.R [hunk 6]
```diff
@@ -10,13 +10,32 @@ library('biomaRt')
...
 }
 library('PLIER')
 
-
 data.dir <- '../data'
 
+args <- commandArgs(trailingOnly=TRUE)
+
+if(length(args) < 2){
+	print('As no value of k was passed in, it will be estimated from the data')
```
Comment (ajlee21) [inline brdnet/run_plier.R:30]: For my own knowledge, I'm curious if these were checks that existed in the PLIER scripts or are these ones that you generated?  If you generated them, how did you come up with these?

**Review by ajlee21 — COMMENTED** [review]
Very nice clean PR.

In the future, it would be helpful if you also uploaded a .py version of your NB for inline comments.  

One minor comment about the nb:  Are you using "outlier" and "classification of disease" synonymously here? 

I think it might help for you to walk me through a bit of the big picture of this project in order to better comment on these scripts.

File: brdnet/create_identity_matrix_df.py (hunk 1 above)
Comment (ben-heil) [reply brdnet/create_identity_matrix_df.py:18]: That's a good idea, I'll definitely update the README. PLIER is a biologically constrained version of PCA, and the identity matrix lets the user train a model on all the genes that could have been used by PLIER.

File: brdnet/evaluate_models.py (hunk 3 above)
Comment (ben-heil) [reply brdnet/evaluate_models.py:117]: Outlier detection methods assume that there are two classes of data, normal data and outliers. When training the methods, the contamination parameter is an estimate of how much of your training data is outliers. Since this problem is a fairly balanced two class classification problem I set the `contamination` parameter to .5 instead of the default .1

File: brdnet/evaluate_models.py (hunk 4 above)
Comment (ben-heil) [reply brdnet/evaluate_models.py:228]: It's the encoding matrix. I forget what it's called in PCA (the loading matrix maybe?), but given the (genes x samples) matrix X, Z is the (genes x latent variables) such that X^T Z gives a lower dimensional representation for the samples. 

I refer to it as the Z matrix throughout the code because the PLIER paper/code refer to it as Z, but I'll give it a more descriptive name if I can think of one.

File: brdnet/model_evaluation_pipeline.sh (hunk 5 above)
Comment (ben-heil) [reply brdnet/model_evaluation_pipeline.sh:12]: These are the number of PCs used by PLIER. The embedding takes a (genes x samples) matrix and reduces it to a (k x samples) matrix

File: brdnet/run_plier.R (hunk 6 above)
Comment (ben-heil) [reply brdnet/run_plier.R:30]: The PLIER vignettes calculate the value of `k` using the function `PLIER::num.pc`. Since I was testing whether I would get better performance by increasing or decreasing the number of PCs, having an optional `k` argument in `run_plier.R` made it possible to create embedding matrices for several values of `k` in parallel

**Comment by ajlee21** [comment]
I have approved this PR after a really helpful discussion with @ben-heil.

Just a few lingering comments based on our discussion (these can be addressed in a future PR):
1.  I think it might be helpful for your future self/users of this repo to add a bit more detail in the readme about the goals of the project, approach, expected output and application.  This is something that can be done as your project develops and you have a more clear understanding of the project.  My readmes are also a work in progress.  But it can be helpful to not get too lost in the weeds as you progress.
2.  Might be helpful to add some documentation on what the dataset it that you're using and how you generated "labels" for them.  This could be something you add to https://nbviewer.jupyter.org/github/ben-heil/brd-net/blob/master/brdnet/download_categorized_data.ipynb or to the readme or both.  Also how you are partitioning your dataset into tuning and validation. 
3.   As discussed, perhaps you might want to consider modifying your baseline definition to a dataset permuted by gene.  This is minor, not sure if its cleaner to have an additional boxplot of the baseline accuracy so you don't have to have a plot with the acc_over_baseline, but that is personal preference.
4.  In terms of metrics I think I would say that it depends on what your question is.  I think accuracy and AUROC makes sense if you want an overall performance for both pos and negatives and your dataset is well balanced as you mentioned.  But if you care more about pos cases perhaps precision and recall make more sense.

**Review by ajlee21 — APPROVED** [review]

**Comment by ben-heil** [comment]
Sounds good! Both the README update and the permuted gene baseline will be big enough projects to be their own PRs, so I'll go ahead and merge this PR.
